clickhouse/client.cpp

```diff
@@ -194,30 +207,55 @@ class Client::Impl {
     std::unique_ptr<InputStream> input_;
     std::unique_ptr<OutputStream> output_;
     std::unique_ptr<SocketBase> socket_;
+    std::shared_ptr<EndpointsIteratorBase> endpoints_iterator;
```

Comment:
Do we really need to store shared pointer to iterator here ? Or we can create it as needed. It looks pretty lightweight

Reply:
For the Round Robin algo, we have to keep the same instance of the object, because otherwise `current_index` will be reset to zero every time the object is created.

Reply:
I agree with Vasily and believe endpoints iteration-related code could be simplified. Having a base class overridden by a specific algorithm is a good general solution, but in case only one algorithm is implemented, I think it's better to have a dead simple `RoundRobinEndpointsIterator endpoints_iterator` without inheritance and smart pointers.

If you still want a general solution, consider using `std::unique_ptr` as we seem to have only one user of this field.